### ISSUE TITLE
DeepGraphInfomax prediction problem

### ISSUE THREAD

**Issue by enmce** [rusty1s/pytorch_geometric#1206]
## ❓ Questions & Help

Hello!
I tried to use DeepGraphInfomax example [here](https://github.com/rusty1s/pytorch_geometric/blob/master/examples/infomax.py) to embed nodes on multiple graphs. Model converges good, but i get weird results then i try to predict embeddings for all graphs in training set: first graph embeddings are making sense and do actually group similar nodes together on UMAP projection, but every other graph after that looks like it`s index was shuffled. To test, i removed first graph on learning phase, and trained on all others - same, first graph (which was not present in training data) have sensible embeddings, all other - completely random on UMAP projection. I removed first graph completely, now second graph shows sensible embeddings, and all others are shuffled. I feel like this may be connected to how data is stored right now, but cant quite understand whats wrong.

I create Data objects like that:
```
from sklearn.preprocessing import OneHotEncoder

data_type_encoder = OneHotEncoder(sparse=False)
data_type_encoder.fit(results['type'].values.reshape(
    -1, 1))

dataset_list = []

for group_id,  group_data in tqdm_notebook(results.groupby(['group'])):
    
    print(group_id, group_data.shape)

    group_adjacency = groups_adj[group]

    graph = nx.Graph()

    for _, data in group_data.iterrows():
        graph.add_node(data['index'])

    for i, near_nodes_idx in enumerate(group_adjacency):
        if i not in graph.nodes():
            continue

        if near_nodes_idx.size == 0:
            continue

        for node_idx in near_nodes_idx:
            if node_idx not in graph.nodes():
                continue

            graph.add_edge(i, node_idx)

    graph = nx.to_scipy_sparse_matrix(graph,
                                      nodelist=group_data['index'].to_list())

    edge_indices = from_scipy_sparse_matrix(graph)[0]

    sample_data = torch.Tensor(
        data_type_encoder.transform(group_data['type'].values.reshape(
            -1, 1)))

    sample_data_object = Data(edge_index=edge_indices,
                              x=sample_data)

    dataset_list.append(sample_data_object)
```
Then is pass this list to dataloader and train, just like in example, except on multiple graphs, so i set cached to False in GCNConv:

```
import torch
import torch.nn as nn
from torch_geometric.nn import GCNConv, DeepGraphInfomax
from torch_geometric.data import Data, DataLoader


dataloader = DataLoader(dataset_list, batch_size=32, shuffle=True)


class Encoder(nn.Module):
    def __init__(self, in_channels, hidden_channels):
        super(Encoder, self).__init__()
        self.conv = GCNConv(in_channels, hidden_channels, cached=False)
        self.conv_2 = GCNConv(hidden_channels, hidden_channels, cached=False)
        self.conv_3 = GCNConv(hidden_channels, hidden_channels, cached=False)
        self.conv_4 = GCNConv(hidden_channels, hidden_channels, cached=False)
        
        self.prelu = nn.PReLU(hidden_channels)
        
    def forward(self, x, edge_index):
        x = self.conv(x, edge_index)
        x = self.conv_2(x, edge_index)
        x = self.conv_3(x, edge_index)
        x = self.conv_4(x, edge_index)
        x = self.prelu(x)

        return x


def corruption(x, edge_index):
    return x[torch.randperm(x.size(0))], edge_index


device = torch.device('cuda:1' if torch.cuda.is_available() else 'cpu')

model = DeepGraphInfomax(
    hidden_channels=64,
    encoder=Encoder(14, 64),
    summary=lambda z, *args, **kwargs: torch.sigmoid(z.mean(dim=0)),
    corruption=corruption).to(device)


optimizer = torch.optim.Adam(model.parameters(), lr=0.001)


def train():
    model.train()
    optimizer.zero_grad()
    
    mean_loss = []
    
    for data in dataloader:
        data = data.to(device)
        pos_z, neg_z, summary = model(data.x, data.edge_index)

        loss = model.loss(pos_z, neg_z, summary)
        loss.backward()
        
        mean_loss.append(loss.item())
        
        optimizer.step()
        
    return np.mean(mean_loss)


for epoch in range(700):
    loss = train()
    print('Epoch: {:03d}, Loss: {:.4f}'.format(epoch, loss))
```

Then i preform prediction like this:

```
group_embeddings = []

dataloader = DataLoader(dataset_list, batch_size=32, shuffle=False)

model.eval()
for data in dataloader:
    data = data.to(device)
    
    pos_z, _, _ = model(data.x, data.edge_index)
    
    node_embeddings = pd.DataFrame(pos_z.detach().cpu().numpy())
    
    group_embeddings.append(node_embeddings)
    
group_embeddings = pd.concat(group_embeddings)
group_embeddings = group_embeddings.reset_index(drop=True)
```

After that, first graph in list is always shows sensible embeddings, all others completely random. What can be the case here?

**Comment by enmce** [comment]
Example, graph is 2D map of nodes with different type, correct embeddings:
![image](https://user-images.githubusercontent.com/13630187/81422167-48f09800-915b-11ea-92e8-b89d3db4d3c6.png)
Now shuffled embeddings, characteristics stripes, like indexes are off:
![image](https://user-images.githubusercontent.com/13630187/81422384-8ce39d00-915b-11ea-9b58-ed9e30514a3a.png)

**Comment by rusty1s** [comment]
That's interesting, as it seems to not generalize to unknown graphs at all. Note that in the paper, the inductive learning setup is quite different to the one applied to transductive tasks. It uses a different GNN operator, and uses a different corruption function (randomly sampled training graphs as negative examples). This might fix your problems.

**Comment by enmce** [comment]
Well the weird thing is, that first graph in example is unseen by network on training. And it produces correct embeddings. I expected overfit and informative embedings on graphs on which network is trained, and random embeddings on test data, here the situation is opposite. It is as if somehow order of graphs on prediction matters, batch shuffles nodes, or i create dataloader incorrectly.

**Comment by enmce** [comment]
Dataset consists of 5 graphs, now i tried to take this situation to extreme, i take only last 5 graph to train on and overfit for 2000 epochs, and then predict embeddings for all of them. Again graph 1 which is first in data list produces correct embeddings, all other, including 5th, on which model is fitted give shuffled embeddings. This reproduces the i remove graph 1 from data list completely, and graph 2 starts to produce correct embeddings, all others look random.

**Comment by enmce** [comment]
Yes, i figured it out! It was indeed error on indexing on my part, when i concatenated the data on prediction phase, probably i have different graph order in original dataframe before groupby operation. In case some one have similar problem with 2d graphs, image of graph with characteristics "stripes" pattern can help to guide your way in solving this bug.

As for model, it generalizes to new graphs quite good, thanks for quick response)

**Comment by rusty1s** [comment]
Awesome :)
